unit/util/code_observer.cpp

```diff
@@ -0,0 +1,109 @@
+/*******************************************************************\
+
+ Module: Code Observer Tests.
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+/// \file
+/// Unit tests for code_observer.h
+
+#include <testing-utils/catch.hpp>
+#include <util/std_types.h>
+#include <util/symbol_table.h>
+#include <util/code_observer.h>
+
+// Utilities
+
+static void add_codet(
+  const exprt& value,
+  const std::string& name,
+  symbol_tablet& symbol_table)
+{
+  symbolt new_symbol;
+  new_symbol.name = name;
+  new_symbol.type = code_typet();
+  new_symbol.value = value;
+  symbol_table.add(new_symbol);
+}
+
+// Observers
+
+class decl_replacert
+{
+public:
+  void replace(codet &code)
+  {
+    REQUIRE(code.id() == ID_code);
+    REQUIRE(code.get_statement() == ID_decl);
+
+    code = codet(ID_skip);
+    replaced += 1;
+  }
+  int replaced;
```

Comment:
You might want a constructor to initialise this to 0?

Reply:
I did this one purpose, to make it clear what the value of replaced is in the test, but will change if you prefer